### ISSUE TITLE
About SQL implementations for keccak (aka sha-3)

### ISSUE THREAD

**Issue by Danilo-Araujo-Silva** [paranoiq/mysql-sha2#1]
Hi @paranoiq,

First, thanks a lot for sharing this code.
This issue is more a question than anything else.

So, I'm searching for an option to use the new SHA-3 or KECCAK-256 / KECCAK-512 hashing functions with Google BigQuery.
Where these new hashing functions are the ones used in the Ethereum (ETH) blockchain for mining new blocks.

Here on the [official keccak website](https://keccak.team/software.html) we can see the references for a lot of different implementations of the algorithms but no SQL yet.

And [here](https://eth.wiki/en/concepts/ethash/ethash) is the Ethereum docs about the ETHASH, the algorithm to mining the blocks and that uses the keccak.

I've seen that we can run js functions on BigQuery, but I think the performance wouldn't be so good, so I was wondering about an SQL implementation. Since you've already done an implementation of the SHA-2 algorithm and considering, for example, this official [python reference](https://github.com/XKCP/XKCP/blob/master/Standalone/CompactFIPS202/Python/CompactFIPS202.py) implementation of the algorithm, do you think it would be too difficult convert these new hashing functions to Standard SQL?

Thanks in advance

**Comment by paranoiq** [comment]
hi. i don't think it is hard to implement, but i have doubts that it would be faster than JS implementation. i do not know the internals of BigQuery, but PL-SQL is usually interpreted, not compiled or JIT-compiled, which makes it pretty slow in computations heavy tasks like this

**Comment by Danilo-Araujo-Silva** [comment]
Thanks a lot @paranoiq for your quick response and interesting comment.

So I did some tests here, where I tested a js function, a wasm function with js and a sql function doing the same (adding two numbers) (since I've found a tutorial of how to run these wasm functions on BigQuery).

It looks like, at least for this case that I've tested, the pure js is performing better than the other ones. Where the worst of them was actually the SQL one. Maybe it is something related with my implementation, but anyway I can see now that maybe proceed with js functions would be the way for me
